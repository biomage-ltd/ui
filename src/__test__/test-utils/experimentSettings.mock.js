import initialState, {
  initialPipelineState,
  initialWorkerState,
} from '../../redux/reducers/experimentSettings/initialState';

const sampleifiedConfig = (sampleIds, configToReplicate) => {
  const result = sampleIds.reduce(
    (acum, sampleId) => {
      // eslint-disable-next-line no-param-reassign
      acum[sampleId] = configToReplicate;
      return acum;
    },
    {},
  );

  return result;
};

const generateProcessingConfigMock = (sampleIds) => ({
  cellSizeDistribution: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      filterSettings: {
        minCellSize: 10800,
        binStep: 200,
      },
    }),
  },
  mitochondrialContent: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      filterSettings: {
        method: 'absolute_threshold',
        methodSettings: {
          absolute_threshold: {
            maxFraction: 0.1,
            binStep: 200,
          },
        },
      },
    }),
  },
  classifier: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      filterSettings: {
        FDR: 0.1,
      },
    }),
  },
  numGenesVsNumUmis: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      filterSettings: {
        regressionType: 'gam',
        regressionTypeSettings: {
          gam: {
            'p.level': 0.00009,
          },
        },
      },
    }),
  },
  doubletScores: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      filterSettings: {
        probabilityThreshold: 0.2,
        binStep: 0.05,
      },
    }),
  },
  dataIntegration: {
    enabled: true,
    dataIntegration: {
      method: 'seuratv4',
      methodSettings: {
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
});

const generateExperimentSettingsMock = (sampleIds) => ({
  ...initialState,
  processing: {
    ...initialState.processing,
    ...generateProcessingConfigMock(sampleIds),
  },
});

export default generateExperimentSettingsMock;

export { initialPipelineState, initialWorkerState };
