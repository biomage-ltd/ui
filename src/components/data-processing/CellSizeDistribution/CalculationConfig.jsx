import React from 'react';
import PropTypes from 'prop-types';
import {
  InputNumber,
  Form,
  Space,
  Tooltip,
} from 'antd';

import {
  InfoCircleOutlined,
} from '@ant-design/icons';
import SliderWithInput from '../../SliderWithInput';

const CellSizeDistributionConfig = (props) => {
  const {
    config, disabled, updateSettings, highestUmi,
  } = props;

  const withinRange = (cellSize) => Math.max(Math.min(cellSize, highestUmi ?? 17000), 0);

  return (
    <>

      <Form.Item disabled label='Minimum #UMIs per cell'>
        <Space>
          <Tooltip title='The cut-off is automatically calculated as the inflection point of the knee plot. The inflection point estimates the boundary between empty drops and droplets that contain cells. The number of UMIs per cell varies depending on cell type. The typical minimum threshold range approx. 500-2000.'>
            <InfoCircleOutlined />
          </Tooltip>
          <InputNumber
            value={config.minCellSize}
            onChange={(value) => {
              updateSettings({ minCellSize: withinRange(value) });
            }}
            onPressEnter={(e) => {
              updateSettings({ minCellSize: withinRange(e.target.value) });
            }}
            placeholder={10800}
            step={100}
            disabled={disabled}
            max={highestUmi ?? 17000}
            min={0}
          />
        </Space>
      </Form.Item>

      <Form.Item label='Bin step'>
        <SliderWithInput
          min={100}
          max={400}
          value={config.binStep}
          onUpdate={(value) => {
            updateSettings({ binStep: value });
          }}
          disabled={disabled}
        />
      </Form.Item>
    </>
  );
};
CellSizeDistributionConfig.propTypes = {
  updateSettings: PropTypes.func.isRequired,
  config: PropTypes.object.isRequired,
  disabled: PropTypes.bool.isRequired,
  highestUmi: PropTypes.number.isRequired,
};

export default CellSizeDistributionConfig;
