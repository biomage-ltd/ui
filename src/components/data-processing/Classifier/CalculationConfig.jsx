import React from 'react';
import PropTypes from 'prop-types';
import {
  Form,
  InputNumber,
  Tooltip,
  Space,
} from 'antd';
import { InfoCircleOutlined } from '@ant-design/icons';

const ClassifierConfig = (props) => {
  const {
    // eslint-disable-next-line no-unused-vars
    config, disabled, updateSettings, plotType,
  } = props;

  return (
    <>
      <Form.Item label='FDR:'>
        <Space direction='horizontal'>
          <Tooltip title='False discovery rate (FDR) is calculated for each barcode by using the ‘emptyDrops’ function (https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html). This distinguishes between droplets containing cells and ambient RNA. The FDR range is [0-1]. The default FDR value is 0.01, where only barcodes with FDR < 0.01 are retained.'>
            <InfoCircleOutlined />
          </Tooltip>
          <InputNumber
            value={config.FDR}
            onChange={(value) => updateSettings({ FDR: value })}
            onPressEnter={(e) => updateSettings({ FDR: e.target.value })}
            placeholder={0.01}
            min={0}
            max={1}
            step={0.01}
            disabled={disabled}
          />
        </Space>
      </Form.Item>
    </>
  );
};

ClassifierConfig.propTypes = {
  updateSettings: PropTypes.func.isRequired,
  config: PropTypes.object.isRequired,
  plotType: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
};

export default ClassifierConfig;
