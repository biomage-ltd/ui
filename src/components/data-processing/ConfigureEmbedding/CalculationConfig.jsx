import React, { useState, useEffect, useCallback } from 'react';
import _ from 'lodash';
import { useDispatch, useSelector } from 'react-redux';
import {
  Collapse, InputNumber, Form, Select, Typography, Tooltip, Slider, Button, Alert,
} from 'antd';
import PropTypes from 'prop-types';
import { QuestionCircleOutlined } from '@ant-design/icons';
import PreloadContent from '../../PreloadContent';

import {
  updateProcessingSettings,
  saveProcessingSettings,
} from '../../../redux/actions/experimentSettings';
import { loadEmbedding } from '../../../redux/actions/embedding';

const { Option } = Select;
const { Panel } = Collapse;
const { Text } = Typography;

const MIN_DIST_TEXT = 'This controls how tightly the embedding is allowed to compress points together. '
  + 'Larger values ensure embedded points are more evenly distributed, while '
  + 'smaller values allow the algorithm to optimise more accurately with regard '
  + 'to local structure. Expected range: 0.001 to 0.5. Default is 0.1.';

const CalculationConfig = (props) => {
  const { experimentId } = props;
  const FILTER_UUID = 'configureEmbedding';

  const dispatch = useDispatch();

  const [changesOutstanding, setChangesOutstanding] = useState(false);

  const data = useSelector((state) => state.experimentSettings.processing[FILTER_UUID]);
  const { method: clusteringMethod } = data?.clusteringSettings || {};
  const { method: embeddingMethod } = data?.embeddingSettings || {};
  const { umap: umapSettings, tsne: tsneSettings } = data?.embeddingSettings.methodSettings || {};
  const { louvain: louvainSettings } = data?.clusteringSettings.methodSettings || {};

  const [resolution, setResolution] = useState(null);
  const [minDistance, setMinDistance] = useState(null);

  useEffect(() => {
    if (!resolution && louvainSettings) {
      setResolution(louvainSettings.resolution);
    }
  }, [louvainSettings]);

  useEffect(() => {
    if (!minDistance && umapSettings) {
      setMinDistance(umapSettings.minimumDistance);
    }
  }, [umapSettings]);

  const dispatchDebounce = useCallback(_.debounce((f) => {
    dispatch(f);
  }, 1500), []);

  const updateSettings = (diff) => {
    if (diff.embeddingSettings) {
      // If this is an embedding change, indicate to user that their changes are not
      // applied until they hit Apply.

      setChangesOutstanding(true);
      dispatch(updateProcessingSettings(
        experimentId,
        FILTER_UUID,
        diff,
      ));
    } else {
      // If it's a clustering change, debounce the save process at 1.5s.
      dispatchDebounce(saveProcessingSettings(experimentId, FILTER_UUID));

      dispatch(updateProcessingSettings(
        experimentId,
        FILTER_UUID,
        diff,
      ));
    }
  };

  // When the Apply button is pressed, remove the warning and save to DynamoDB.
  const applyEmbeddingSettings = () => {
    setChangesOutstanding(false);
    dispatch(saveProcessingSettings(experimentId, FILTER_UUID));
    dispatch(loadEmbedding(experimentId, embeddingMethod));
  };

  const renderUMAPSettings = () => (
    <>
      <Form.Item>
        <Text strong>Settings for UMAP:</Text>
      </Form.Item>
      <Form.Item label={(
        <span>
          Minimum distance&nbsp;
          <Tooltip title={MIN_DIST_TEXT}>
            <QuestionCircleOutlined />
          </Tooltip>
        </span>
      )}
      >
        <InputNumber
          value={minDistance}
          min={0}
          step={0.1}
          onChange={(value) => setMinDistance(value)}
          onStep={(value) => updateSettings(
            {
              embeddingSettings:
                { methodSettings: { umap: { minimumDistance: parseFloat(value) } } },
            },
          )}
          onPressEnter={(e) => e.preventDefault()}
          onBlur={(e) => {
            updateSettings(
              {
                embeddingSettings:
                  { methodSettings: { umap: { minimumDistance: parseFloat(e.target.value) } } },
              },
            );
          }}
        />
      </Form.Item>
      <Form.Item label='Distance metric:'>
        <Select
          value={umapSettings.distanceMetric}
          onChange={(value) => updateSettings(
            { embeddingSettings: { methodSettings: { umap: { distanceMetric: value } } } },
          )}
        >
          <Option value='euclidean'><Tooltip style={{ width: '100%' }} title='que tal'>Euclidean</Tooltip></Option>
          <Option value='cosine' disabled>Cosine</Option>
        </Select>
      </Form.Item>
    </>
  );

  const renderTSNESettings = () => (
    <>
      <Form.Item>
        <Text strong>Settings for t-SNE:</Text>
      </Form.Item>
      <Form.Item label='Perplexity:'>
        <InputNumber
          value={tsneSettings.perplexity}
          min={5}
          onInput={(e) => updateSettings(
            {
              embeddingSettings:
                { methodSettings: { tsne: { perplexity: parseFloat(e.target.value) } } },
            },
          )}
          onStep={(value) => updateSettings(
            {
              embeddingSettings:
                { methodSettings: { tsne: { perplexity: parseFloat(value) } } },
            },
          )}
        />
      </Form.Item>
      <Form.Item label='Learning rate:'>
        <InputNumber
          value={tsneSettings.learningRate}
          min={10}
          max={1000}
          step={10}
          onInput={(e) => updateSettings(
            {
              embeddingSettings:
                { methodSettings: { tsne: { learningRate: parseFloat(e.target.value) } } },
            },
          )}
          onStep={(value) => updateSettings(
            {
              embeddingSettings:
                { methodSettings: { tsne: { learningRate: parseFloat(value) } } },
            },
          )}

        />
      </Form.Item>
    </>
  );

  if (!data) {
    return <PreloadContent />;
  }

  return (
    <Collapse defaultActiveKey={['embedding-settings', 'clustering-settings']}>
      <Panel header='Embedding settings' key='embedding-settings' collapsible={changesOutstanding ? 'disabled' : 'header'}>
        <Form size='small'>
          {changesOutstanding && (
            <Form.Item>
              <Alert message='Your changes are not yet applied. To update the plots, click Apply.' type='warning' showIcon />
            </Form.Item>
          )}
          <Form.Item label='Method:'>
            <Select
              value={embeddingMethod}
              onChange={(value) => updateSettings(
                { embeddingSettings: { method: value } },
              )}
            >
              <Option value='umap'>UMAP</Option>
              <Option value='tsne'>t-SNE</Option>
            </Select>
          </Form.Item>
          {embeddingMethod === 'umap' && renderUMAPSettings()}
          {embeddingMethod === 'tsne' && renderTSNESettings()}
          <Form.Item>
            <Tooltip title={!changesOutstanding ? 'No outstanding changes' : ''}>
              <Button type='primary' htmlType='submit' disabled={!changesOutstanding} onClick={applyEmbeddingSettings}>Apply</Button>
            </Tooltip>
          </Form.Item>
        </Form>
      </Panel>
      <Panel header='Clustering settings' key='clustering-settings'>
        <Form size='small'>
          <Form.Item label='Clustering method:'>
            <Select
              value={clusteringMethod}
              onChange={(value) => updateSettings(
                { clusteringSettings: { method: value } },
              )}
            >
              <Option value='louvain'>Louvain</Option>
              <Option value='leiden' disabled>Leiden</Option>
              <Option value='slm' disabled>SLM</Option>
            </Select>
          </Form.Item>
          <Form.Item label='Resolution'>
            <Slider
              value={resolution}
              min={0}
              max={2}
              step={0.1}
              onChange={(value) => {
                setResolution(value);
              }}
              onAfterChange={(value) => {
                updateSettings({
                  clusteringSettings: {
                    methodSettings: {
                      louvain: { resolution: value },
                    },
                  },
                });
              }}
            />
          </Form.Item>
        </Form>
      </Panel>
    </Collapse>
  );
};

CalculationConfig.propTypes = {
  experimentId: PropTypes.string.isRequired,
};

export default CalculationConfig;
