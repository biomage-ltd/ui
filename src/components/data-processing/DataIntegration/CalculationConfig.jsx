import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import {
  Collapse,
  Space,
  Select,
  InputNumber,
  Form,
  Checkbox,
  Tooltip,
  Button,
  Typography,
  Alert,
  Row,
  Col,
} from 'antd';

import {
  QuestionCircleOutlined,
} from '@ant-design/icons';

import _ from 'lodash';

import SeuratV4Options from './SeuratV4Options';

import { updateProcessingSettings } from '../../../redux/actions/experimentSettings';
import generateDataProcessingPlotUuid from '../../../utils/generateDataProcessingPlotUuid';

const { Option } = Select;
const { Text } = Typography;
const { Panel } = Collapse;

const CalculationConfig = (props) => {
  const {
    experimentId, onPipelineRun, disabled,
  } = props;
  const FILTER_UUID = 'dataIntegration';

  const dispatch = useDispatch();
  const { dataIntegration, dimensionalityReduction } = useSelector(
    (state) => state.experimentSettings.processing.dataIntegration,
  );

  const elbowPlotUuid = generateDataProcessingPlotUuid(null, FILTER_UUID, 1);
  const data = useSelector((state) => state.componentConfig[elbowPlotUuid]?.plotData);

  const methods = [
    {
      value: 'seuratv4',
      text: 'Seurat v4',
      disabled: false,
    },
    {
      value: 'seuratv3',
      text: 'Seurat v3',
      disabled: true,
    },
    {
      value: 'harmony',
      text: 'Harmony',
      disabled: true,
    },
    {
      value: 'conos',
      text: 'Conos',
      disabled: true,
    },
    {
      value: 'liger',
      text: 'Liger',
      disabled: true,
    },
    {
      value: 'fastMNN',
      text: 'Fast MNN',
      disabled: true,
    },
  ];

  const [numPCs, setNumPCs] = useState(dimensionalityReduction.numPCs);
  const [changesOutstanding, setChangesOutstanding] = useState(false);

  const updateSettings = (diff) => {
    setChangesOutstanding(true);
    dispatch(updateProcessingSettings(
      experimentId,
      FILTER_UUID,
      diff,
    ));
  };

  const runWithCurrentDataIntegrationSettings = () => {
    setChangesOutstanding(false);
    onPipelineRun();
  };

  const methodOptions = {
    seuratv4: () => (
      <SeuratV4Options
        config={dataIntegration.methodSettings.seuratv4}
        onUpdate={updateSettings}
        onChange={() => setChangesOutstanding(true)}
        disabled={disabled}
      />
    ),
  };

  const roundedVariationExplained = () => {
    const variationExplained = data?.length
      ? data.slice(0, dimensionalityReduction.numPCs).reduce(
        (acum, current) => acum + current.percentVariance, 0,
      ) : 0;
    const roundingPrecision = 2;

    return _.round(variationExplained * 100, roundingPrecision);
  };

  return (
    <Collapse defaultActiveKey={['data-integration']}>
      <Panel header='Data Integration' key='data-integration'>
        <Space direction='vertical' style={{ width: '100%' }} />
        <Form size='small'>
          {changesOutstanding && (
            <Form.Item>
              <Alert
                message='Your changes are not yet applied. To rerun data integration, click Apply.'
                type='warning'
                showIcon
              />
            </Form.Item>
          )}
          <Form.Item>
            <Text>
              <strong style={{ marginRight: '0.5rem' }}>Data integration settings:</strong>
              <Tooltip title='Integration of multiple samples corrects for batch effect. These methods identify shared cell states that are present across different datasets, even if they were collected from different individuals, experimental conditions, technologies, or even species. The user selects the integration method and sets the controls, as appropriate. The latest Seurat method is selected as default.'>
                <QuestionCircleOutlined />
              </Tooltip>
            </Text>
          </Form.Item>
          <div style={{ paddingLeft: '1rem' }}>
            <Form.Item
              label='Method:'
            >
              <Select
                value={dataIntegration.method}
                onChange={(val) => updateSettings({ dataIntegration: { method: val } })}
                disabled={disabled}
              >
                {
                  methods.map((el) => (
                    <Option key={el.text} value={el.value} disabled={el.disabled}>{el.text}</Option>
                  ))
                }
              </Select>
            </Form.Item>

            {
              methodOptions[dataIntegration.method]()
            }

          </div>
          <Form.Item>
            <Text>
              <strong style={{ marginRight: '0.5rem' }}>Dimensionality reduction settings:</strong>
              <Tooltip title='Dimensionality reduction is necessary to summarise and visualise single cell RNA-seq data. The most common method is Principal Component Analysis. The user sets the number of Principal Components (PCs). This is the number that explains the majority of the variation within the dataset (ideally >90%), and is typically set between 5 and 30.'>
                <QuestionCircleOutlined />
              </Tooltip>
            </Text>
          </Form.Item>
          <div style={{ paddingLeft: '1rem' }}>
            <Form.Item label='Number of Principal Components'>
              <InputNumber
                value={numPCs}
                max={data?.length || 100}
                min={0}
                onChange={(value) => {
                  setChangesOutstanding(true);
                  setNumPCs(value);
                }}
                onPressEnter={(e) => e.preventDefault()}
                onStep={(value) => updateSettings({ dimensionalityReduction: { numPCs: value } })}
                onBlur={(e) => updateSettings(
                  { dimensionalityReduction: { numPCs: parseInt(e.target.value, 0) } },
                )}
                disabled={disabled}
              />
            </Form.Item>
            <Form.Item label='% variation explained'>
              <InputNumber
                value={roundedVariationExplained()}
                disabled={disabled}
                readOnly
              />
            </Form.Item>
            <Form.Item label='Exclude genes categories'>
              <Tooltip title='Normalization can be biased by certain gene categories such the ones listed here.
              Checking them will ignore those categories.
              For example, cell cycle genes should be removed if sampling timepoints occured throughout the day. 
              Those genes can otherwise introduces within-cell-type heterogeneity that can obscure the differences 
              in expression between cell types.
              This is not implemented yet'>
                <QuestionCircleOutlined />
              </Tooltip>
              <Checkbox.Group
                onChange={(val) => updateSettings(
                  { dimensionalityReduction: { excludeGeneCategories: val } },
                )}
                value={dimensionalityReduction.excludeGeneCategories}
                disabled={disabled}
              >
                <Space direction='vertical'>
                  <Checkbox value='ribosomal'>ribosomal</Checkbox>
                  <Checkbox value='mitochondrial'>mitochondrial</Checkbox>
                  <Checkbox value='cellCycle'>cell cycle</Checkbox>
                </Space>
              </Checkbox.Group>
            </Form.Item>
            <Form.Item label={(
              <span>
                Method&nbsp;
                <Tooltip overlay={(
                  <span>
                    To integrate data, dimensional reduction is performed to find so called "anchors".
                    cross-dataset pairs of cells that are in a matched biological state (‘anchors’), can are both to correct for technical
                    differences between datasets
                    (i.e. batch effect correction), and to perform comparative scRNA-seq analysis across experimental conditions.
                    CCA is well-suitedn cell types are conserved, but there are very substantial differences
                    in gene expression across experiments.
                    However, CCA-based integration may also lead to overcorrection, especially when a large proportion of cells are
                    non-overlapping across datasets.

                    RPCA-based integration runs significantly faster, and also represents a more conservative approach where
                    cells in different biological states are less likely to ‘align’ after integration.
                    More info
                    <a
                      href='https://satijalab.org/seurat/articles/integration_rpca.html'
                      target='_blank'
                      rel='noreferrer'
                    >
                      {' '}
                      <code>here</code>
                    </a>
                  </span>
                )}
                >
                  <QuestionCircleOutlined />
                </Tooltip>
              </span>
            )}
            >
              <Select
                value={dimensionalityReduction.method}
                onChange={(val) => updateSettings({ dimensionalityReduction: { method: val } })}
                disabled={disabled}
              >
                <Option key='rpca' value='rpca'>Reciprocal PCA (RPCA)</Option>
                <Option key='cca' value='cca'>Cannonical Correlation Analysis (CCA)</Option>
              </Select>

            </Form.Item>
            <Form.Item>
              <Row>
                <Col span={6}>
                  <Tooltip title={!changesOutstanding ? 'No outstanding changes' : ''}>
                    <Button
                      size='small'
                      type='primary'
                      htmlType='submit'
                      disabled={!changesOutstanding}
                      onClick={runWithCurrentDataIntegrationSettings}
                    >
                      Run
                    </Button>
                  </Tooltip>
                </Col>
              </Row>
            </Form.Item>
          </div>
        </Form>
      </Panel>
    </Collapse >
  );
};

CalculationConfig.propTypes = {
  experimentId: PropTypes.string.isRequired,
  onPipelineRun: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
};

CalculationConfig.defaultProps = {
  disabled: false,
};

export default CalculationConfig;
