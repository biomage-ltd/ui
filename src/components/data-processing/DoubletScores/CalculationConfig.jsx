import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Space,
  Slider,
  Form,
  Button,
  Alert,
  Radio,
} from 'antd';

import _ from 'lodash';

import BandwidthOrBinstep from '../ReadAlignment/PlotStyleMisc';

import { updateProcessingSettings } from '../../../redux/actions/experimentSettings';

const CalculationConfig = (props) => {
  const {
    experimentId, sampleId, sampleIds,
  } = props;

  const config = useSelector(
    (state) => state.experimentSettings.processing.doubletScores[sampleId]?.filterSettings
      || state.experimentSettings.processing.doubletScores.filterSettings,
  );

  const FILTER_UUID = 'doubletScores';

  const dispatch = useDispatch();

  const [displayIndividualChangesWarning, setDisplayIndividualChangesWarning] = useState(false);

  const updateAllSettings = () => {
    setDisplayIndividualChangesWarning(false);

    const newConfig = {};
    sampleIds.forEach((currentSampleId) => {
      newConfig[currentSampleId] = { filterSettings: config };
    });

    dispatch(updateProcessingSettings(
      experimentId,
      FILTER_UUID,
      newConfig,
    ));
  };

  const updateSettings = (diff) => {
    const newConfig = _.cloneDeep(config);
    _.merge(newConfig, diff);

    const sampleSpecificDiff = { [sampleId]: { filterSettings: newConfig } };

    setDisplayIndividualChangesWarning(true);
    dispatch(updateProcessingSettings(
      experimentId,
      FILTER_UUID,
      sampleSpecificDiff,
    ));
  };

  const filtering = false;

  return (
    <>
      <Space direction='vertical' style={{ width: '100%' }} />
      {displayIndividualChangesWarning && (
        <Form.Item>
          <Alert
            message='Your changes are only applied to this sample. To apply it to all other samples, click Apply to all samples.'
            type='warning'
            showIcon
          />
        </Form.Item>
      )}
      <Radio.Group defaultValue={1} style={{ marginTop: '5px', marginBottom: '30px' }}>
        <Radio value={1}>
          Automatic
        </Radio>
        <Radio value={2}>
          Manual
        </Radio>
      </Radio.Group>
      <Form.Item label='Probability threshold'>
        <Slider
          collapsible={!filtering ? 'disabled' : 'header'}
          value={config.probabilityThreshold}
          min={0}
          max={1}
          onChange={(val) => updateSettings({ probabilityThreshold: val })}
          step={0.05}
        />
      </Form.Item>
      <BandwidthOrBinstep
        config={config}
        onUpdate={updateSettings}
        type='bin step'
      />
      <Button onClick={updateAllSettings}>Apply to all samples</Button>
    </>
  );
};

CalculationConfig.propTypes = {
  experimentId: PropTypes.string.isRequired,
  sampleId: PropTypes.string.isRequired,
  sampleIds: PropTypes.array.isRequired,
};

export default CalculationConfig;
