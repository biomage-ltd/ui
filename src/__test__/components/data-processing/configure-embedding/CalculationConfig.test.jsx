import React from 'react';
import _ from 'lodash';
import {
  Select, Form, Alert, Button,
} from 'antd';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import preloadAll from 'jest-next-dynamic';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';

import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import CalculationConfig from '../../../../components/data-processing/ConfigureEmbedding/CalculationConfig';
import { initialEmbeddingState } from '../../../../redux/reducers/embeddings/initialState';
import generateExperimentSettingsMock from '../../../test-utils/experimentSettings.mock';

jest.mock('localforage');
enableFetchMocks();
const mockStore = configureStore([thunk]);

const initialExperimentState = generateExperimentSettingsMock([]);

describe('Data Processing CalculationConfig', () => {
  const storeState = {
    embeddings: initialEmbeddingState,
    experimentSettings: {
      ...initialExperimentState,
      backendStatus: {
        status: {
          pipeline: {
            startDate: '2020-01-01T00:00:00',
          },
        },
      },
    },
  };

  const mockOnConfigChange = jest.fn(() => {});
  const mockOnPipelineRun = jest.fn(() => {});

  configure({ adapter: new Adapter() });

  beforeEach(async () => {
    await preloadAll();

    const response = new Response(
      JSON.stringify(
        {
          processingConfig:
            { ...initialExperimentState.processing },
        },
      ),
    );

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValue(response);
  });

  afterEach(() => {
    jest.clearAllMocks();
    mockOnConfigChange.mockClear();
    mockOnPipelineRun.mockClear();
  });

  it('renders correctly when nothing is loaded', () => {
    const store = mockStore({
      embeddings: {},
      experimentSettings: {
        ...initialExperimentState,
        processing: {
          ...initialExperimentState.processing,
          configureEmbedding: null,
        },
      },
    });

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    const preloadContent = component.find('PreloadContent');

    // There should be a spinner for loading state.
    expect(preloadContent.length).toEqual(1);
  });

  it('renders correctly when the data is in the store', () => {
    const store = mockStore(storeState);

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    // There should no spinner anymore.
    const preloadContent = component.find('PreloadContent');
    expect(preloadContent.length).toEqual(0);

    // There should be a form loaded.
    const form = component.find(Form);
    expect(form.length).toBeGreaterThan(0);
  });

  it('A changed setting should show an alert', () => {
    const changedStepStoreState = _.cloneDeep(storeState);
    changedStepStoreState.experimentSettings.processing.meta.changedQCFilters = new Set(['configureEmbedding']);
    const store = mockStore(changedStepStoreState);

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    // The alert should show up.
    const alert = component.find(Alert);
    expect(alert.length).toEqual(1);

    // The button should be enabled.
    const button = component.find(Button);
    expect(button.at(0).getElement().props.disabled).toEqual(false);
  });

  it('a changed setting should trigger an onConfigChange callback', () => {
    const store = mockStore(storeState);

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    // The Apply button should be disabled.
    const button = component.find(Button);

    expect(button.at(0).getElement().props.disabled).toEqual(true);

    // Switching the embedding type...
    act(() => { component.find(Select).at(0).getElement().props.onChange('tsne'); });

    component.update();

    expect(mockOnConfigChange).toHaveBeenCalledTimes(1);
  });

  it('clicking on button triggers save action and reloading of plot data', async () => {
    const changedStepStoreState = _.cloneDeep(storeState);
    changedStepStoreState.experimentSettings.processing.meta.changedQCFilters = new Set(['configureEmbedding']);
    const store = mockStore(changedStepStoreState);

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    // Clicking the Apply button.
    const button = component.find(Button);

    button.simulate('click', {});

    // Should trigger onPipelineRun
    expect(mockOnPipelineRun).toHaveBeenCalledTimes(1);
  });

  it('Clicking run with other filters changed triggers the pipeline', async () => {
    const store = mockStore({
      ...storeState,
      experimentSettings: {
        ...storeState.experimentSettings,
        processing: {
          ...storeState.experimentSettings.processing,
          meta: {
            changedQCFilters: new Set(['filter1', 'awesomeFilter']),
          },
        },
      },
    });

    const component = mount(
      <Provider store={store}>
        <CalculationConfig
          experimentId='1234'
          width={50}
          height={50}
          onPipelineRun={mockOnPipelineRun}
          onConfigChange={mockOnConfigChange}
        />
      </Provider>,
    );

    act(() => { component.find(Select).at(0).getElement().props.onChange('tsne'); });
    component.update();

    const runButton = component.find(Button);
    runButton.simulate('click');
    expect(mockOnPipelineRun).toBeCalledTimes(1);
  });
});
