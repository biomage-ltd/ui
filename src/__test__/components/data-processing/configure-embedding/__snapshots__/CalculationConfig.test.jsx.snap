// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Processing CalculationConfig clicking on button triggers save action and reloading of plot data 1`] = `
Array [
  Object {
    "payload": Object {
      "configChange": Object {
        "embeddingSettings": Object {
          "method": "tsne",
          "methodSettings": Object {
            "tsne": Object {
              "learningRate": 200,
              "perplexity": 30,
            },
            "umap": Object {
              "distanceMetric": "euclidean",
              "minimumDistance": 0.1,
            },
          },
        },
      },
      "step": "configureEmbedding",
    },
    "type": "experimentSettings/updateProcessing",
  },
  Object {
    "payload": Object {
      "embeddingType": "umap",
      "experimentId": "1234",
    },
    "type": "embeddings/loading",
  },
]
`;
